ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,325 @@
+    
+    s_entries = Memory_AllocArray<VnodeCacheEntry>(s_entriesCapacity);
+    if (nullptr == s_entries)
+    {
+        s_entriesCapacity = 0;
+        return KERN_RESOURCE_SHORTAGE;
+    }
+    
+    memset(s_entries, 0, s_entriesCapacity * sizeof(VnodeCacheEntry));
+    
+    PerfTracing_RecordSample(PrjFSPerfCounter_CacheCapacity, 0, s_entriesCapacity);
+    
+    return KERN_SUCCESS;
+}
+
+kern_return_t VnodeCache_Cleanup()
+{
+    if (nullptr != s_entries)
+    {
+        Memory_FreeArray<VnodeCacheEntry>(s_entries, s_entriesCapacity);
+        s_entries = nullptr;
+        s_entriesCapacity = 0;
+    }
+    
+    if (RWLock_IsValid(s_entriesLock))
+    {
+        RWLock_FreeMemory(&s_entriesLock);
+        return KERN_SUCCESS;
+    }
+    
+    return KERN_FAILURE;
+}
+
+VirtualizationRootHandle VnodeCache_FindRootForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheHitCounter,
+    PrjFSPerfCounter cacheMissCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    vfs_context_t _Nonnull context)
+{
+    VirtualizationRootHandle rootHandle = RootHandle_None;
+    uintptr_t vnodeHashIndex = ComputeVnodeHashIndex(vnode);
+    uint32_t vnodeVid = vnode_vid(vnode);
+    
+    if (TryGetVnodeRootFromCache(vnode, vnodeHashIndex, vnodeVid, rootHandle))
+    {
+        perfTracer->IncrementCount(cacheHitCounter, true /*ignoreSampling*/);
+        return rootHandle;
+    }
+    
+    perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+    LookupVnodeRootAndUpdateCache(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        context,
+        vnode,
+        vnodeHashIndex,
+        vnodeVid,
+        false, // forceRefreshEntry
+        rootHandle);
+    
+    return rootHandle;
+}
+
+VirtualizationRootHandle VnodeCache_RefreshRootForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheHitCounter,
+    PrjFSPerfCounter cacheMissCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    vfs_context_t _Nonnull context)
+{
+    VirtualizationRootHandle rootHandle = RootHandle_None;
+    uintptr_t vnodeHashIndex = ComputeVnodeHashIndex(vnode);
+    uint32_t vnodeVid = vnode_vid(vnode);
+    
+    perfTracer->IncrementCount(cacheMissCounter, true /*ignoreSampling*/);
+    LookupVnodeRootAndUpdateCache(
+        perfTracer,
+        cacheMissFallbackFunctionCounter,
+        cacheMissFallbackFunctionInnerLoopCounter,
+        context,
+        vnode,
+        vnodeHashIndex,
+        vnodeVid,
+        true, // forceRefreshEntry
+        rootHandle);
+    
+    return rootHandle;
+}
+
+void VnodeCache_InvalidateCache(PerfTracer* _Nullable perfTracer)
+{
+    if (nullptr != perfTracer)
```

Comment:
When would `perfTracer` be `nullptr`? Only in unit tests? If so, I'd force this to be non-null and just pass in a dummy in the tests.

Reply:
Great cache, I'll see if we still need this, but I suspect it's something that could have been dropped in an an earlier refactoring and was missed.